File: dataset/CoT/OPENAI_gpt-5-2025-08-07/Task187.java
import java.util.*;
import java.util.concurrent.*;
import java.util.concurrent.atomic.AtomicInteger;

public class Task187 {

    // Synchronization class for H2O formation
    static class H2O {
        private final Semaphore hSem = new Semaphore(2, true);
        private final Semaphore oSem = new Semaphore(1, true);
        private final CyclicBarrier barrier = new CyclicBarrier(3);

        public void hydrogen(Runnable releaseHydrogen) {
            try {
                hSem.acquire();
                try {
                    releaseHydrogen.run();
                    barrier.await();
                } finally {
                    hSem.release();
                }
            } catch (InterruptedException ie) {
                Thread.currentThread().interrupt();
            } catch (BrokenBarrierException bbe) {
                // Barrier broken, allow graceful exit
            }
        }

        public void oxygen(Runnable releaseOxygen) {
            try {
                oSem.acquire();
                try {
                    releaseOxygen.run();
                    barrier.await();
                } finally {
                    oSem.release();
                }
            } catch (InterruptedException ie) {
                Thread.currentThread().interrupt();
            } catch (BrokenBarrierException bbe) {
                // Barrier broken, allow graceful exit
            }
        }
    }

    // Runs the water formation concurrently and returns the resulting sequence
    public static String runWater(String water) {
        if (water == null) return "";
        final StringBuffer out = new StringBuffer(water.length());
        H2O h2o = new H2O();
        List<Thread> threads = new ArrayList<>(water.length());

        Runnable releaseHydrogen = () -> out.append('H');
        Runnable releaseOxygen = () -> out.append('O');

        for (int i = 0; i < water.length(); i++) {
            char c = water.charAt(i);
            if (c == 'H') {
                Thread t = new Thread(() -> h2o.hydrogen(releaseHydrogen));
                threads.add(t);
                t.start();
            } else if (c == 'O') {
                Thread t = new Thread(() -> h2o.oxygen(releaseOxygen));
                threads.add(t);
                t.start();
            } else {
                // ignore invalid characters
            }
        }

        for (Thread t : threads) {
            try {
                t.join();
            } catch (InterruptedException ie) {
                Thread.currentThread().interrupt();
            }
        }
        return out.toString();
    }

    // Validate output: each group of 3 must contain exactly 2 H and 1 O
    public static boolean validateWater(String out) {
        if (out == null || out.length() % 3 != 0) return false;
        for (int i = 0; i < out.length(); i += 3) {
            int h = 0, o = 0;
            for (int j = i; j < i + 3; j++) {
                char c = out.charAt(j);
                if (c == 'H') h++;
                else if (c == 'O') o++;
                else return false;
            }
            if (!(h == 2 && o == 1)) return false;
        }
        return true;
    }

    public static void main(String[] args) {
        String[] tests = new String[] {
            "HOH",
            "OOHHHH",
            "HHOHHO",
            "OHHHOH",
            "HHOOHH"
        };
        for (String in : tests) {
            String out = runWater(in);
            System.out.println("Input:  " + in);
            System.out.println("Output: " + out + " | valid=" + validateWater(out));
            System.out.println("---");
        }
    }
}

File: dataset/CoT - Copy/GEMINI_gemini-2.5-pro/Task187.cpp
#include <iostream>
#include <string>
#include <vector>
#include <thread>
#include <functional>
#include <semaphore>
#include <barrier>
#include <mutex>

class H2O {
private:
    std::counting_semaphore<2> hSem;
    std::counting_semaphore<1> oSem;
    std::barrier<> barrier;
    
public:
    H2O() : hSem(2), oSem(1), barrier(3, [&]() noexcept {
        hSem.release(2);
        oSem.release(1);
    }) {}

    void hydrogen(std::function<void()> releaseHydrogen) {
        hSem.acquire();
        // releaseHydrogen() outputs "H".
        releaseHydrogen();
        barrier.arrive_and_wait();
    }

    void oxygen(std::function<void()> releaseOxygen) {
        oSem.acquire();
        // releaseOxygen() outputs "O".
        releaseOxygen();
        barrier.arrive_and_wait();
    }
};

class Task187 {
public:
    static std::string runWaterSimulation(const std::string& water) {
        H2O h2o;
        std::string result = "";
        std::mutex resultMutex;
        std::vector<std::thread> threads;

        for (char atom : water) {
            if (atom == 'H') {
                threads.emplace_back(&H2O::hydrogen, &h2o, [&]() {
                    std::lock_guard<std::mutex> lock(resultMutex);
                    result += 'H';
                });
            } else if (atom == 'O') {
                threads.emplace_back(&H2O::oxygen, &h2o, [&]() {
                    std::lock_guard<std::mutex> lock(resultMutex);
                    result += 'O';
                });
            }
        }

        for (auto& t : threads) {
            if (t.joinable()) {
                t.join();
            }
        }

        return result;
    }

    static void main() {
        std::vector<std::string> testCases = {
            "HOH",
            "OOHHHH",
            "HHOHHO",
            "HHHHHHOOO",
            "OHHOHHOHH"
        };

        for (int i = 0; i < testCases.size(); ++i) {
            std::cout << "Test Case " << (i + 1) << ":" << std::endl;
            std::cout << "Input: " << testCases[i] << std::endl;
            std::string output = runWaterSimulation(testCases[i]);
            std::cout << "Output: " << output << std::endl;
            std::cout << "--------------------" << std::endl;
        }
    }
};

int main() {
    Task187::main();
    return 0;
}